internal/types/document_validation.go

```diff
@@ -50,6 +53,11 @@ func (d *Document) ValidateData() error {
 		if strings.Contains(key, "$") {
 			return newValidationError(fmt.Errorf("invalid key: %q (key must not contain $)", key))
 		}
+
+		if _, ok := duplicateChecker[key]; ok {
+			return newValidationError(fmt.Errorf("invalid key: %q (duplicate keys are not allowed)", key))
```

Comment:
Shall we add validation test for this scenario.

Reply:
added a unit test and integration test (skipped for another reason for now)